File: myLeetCode/Binary_Tree/Traversal mode of Binary tree/recursion/145. Binary Tree Postorder Traversal.cpp
// 145. Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/

#include "../../BinaryTreeUtils.hpp"
#include "../../printUtils.hpp"

template <typename T>
void postorderTraversal(TreeNode<T>* root, std::vector<T>& result) {
    if (root == nullptr) return;
    postorderTraversal(root->left, result); // 递归遍历左子树
    postorderTraversal(root->right, result); // 递归遍历右子树
    result.push_back(root->val); // 将当前节点的值加入结果数组
}

template <typename T>
std::vector<T> postorderTraversal(TreeNode<T>* root) {
    std::vector<T> result;
    postorderTraversal(root, result);
    return result;
}

int main() {
    TreeNode<int>* root = buildTree({1, 2, 3, 4, -1, 5, 6}, -1);
    std::cout << "Postorder Traversal: ";
    printArray(postorderTraversal(root));
    deleteTree(root);
    
    return 0;
}

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/recursion/145. Binary Tree Postorder Traversal.py
from typing import Optional
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        return self.postorderTraversal(root.left) + self.postorderTraversal(root.right) + [root.val] if root else []

sol = Solution()

# 测试用例
root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
print(sol.postorderTraversal(root))  # 应输出 [3, 2, 1]
